User Roles Blog (#1)
drumusician/student_manager: blog/user_roles.md

# Real World Phoenix
## User Roles

Welcome back! In the last episode of this series we implemented user authentication.
Now let's see how we can implement user authorization by implementing different user roles into our system.

We used Pow! for our authentication and that library also has a guide for implementing a user roles system.

> Let's see how far that get's us and I'd like to actually take it a bit further and implement different

Reviewer: ```suggestion
Let's see how far that gets us and I'd like to actually take it a bit further and implement different
```

sign up flows for these different user roles.

But as always, we shouldn't get ahead of ourselves and make it too complicated too soon, so we'll first

> implement basic user roles. We are going to create these users in a simple manner by just adding a selection

Reviewer: I'm not sure I understand what you're trying to say with:

> We are going to create these users in a simple manner by just adding a selection
before the registration form that will set this field for the user

Author: yeah that wasn't very clear fixing it

before the registration form that will set this field for the user. I already know I want to be able to define
multple roles in the future so I'll take that into account when creating the field I'm going to use, but we'll
not complicate things too much so 1 role can be selected when signing up.

### Migration and schema

> We'll need to store the user roles, so we'll add a migration to store a list of users in the users table:

Reviewer: ```suggestion
We'll need to store the user roles, so we'll add a migration to store a list of roles in the users table:
```

```elixir
defmodule StudentManager.Repo.Migrations.AddRolesToUsers do
  use Ecto.Migration

  def change do
    alter table(:users) do
      add :roles, {:array, :string}, default: ["student"]
    end
  end
end
```

In this migration I'm taking advantage of the Postgres Array type to store the list of users in the db. It is
of course also possible to store the roles in a different table and at those as an association, but for now
this will do and we can just validate the allowed role values in the Changeset we'll use to store and update
this user role info.

We'll also need to add this new field to our User struct to be able to use it in our application. Here I have also added the validation of the user role values. 

```elixir
defmodule StudentManager.Accounts.User do
  use Ecto.Schema
  use Pow.Ecto.Schema

  schema "users" do
    field(:roles, {:array, :string}, default: ["student"])

    pow_user_fields()

    timestamps()
  end

  def changeset_role(user_or_changeset, attrs) do
    user_or_changeset

>     |> Changeset.cast(attrs, [:roles])

Reviewer: I think this is a little dangerous, as it quite easily could lead to users changing their own role if you don't white/black list attributes somewhere higher up.

We've gone a bit back and forth on this on other projects and I still haven't found the perfect solution, but for now I prefer to lock it down already at this level. For example with explicit changeset functions:

```ex
def student_role_changeset(user) do
  change(user, %{roles: ["student"]})
end

def teacher_role_changeset(user) do
  change(user, %{roles: ["teacher"]})
end
```

WDYT?

Author: yes, I was actually doing something similar already, but I did leave in one line that exposes this vulnerability indeed. So that shouldn't be there. See below

```elixir
  def changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> pow_changeset(attrs)
    |> changeset_role(attrs) # <= **this is dangerous!**
  end

  def teacher_changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> changeset(attrs)
    |> changeset_role(%{roles: ["teacher"]})
  end

  def student_changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> changeset(attrs)
    |> changeset_role(%{roles: ["student"]})
  end

  def changeset_role(user_or_changeset, attrs) do
    user_or_changeset
    |> cast(attrs, [:roles])
    |> validate_subset(:roles, ~w(student admin teacher parent))
  end
```

However, your solution with the `change/2` function is actually much more succinct. The benefit of the `changeset_role/2`, could be to validate things entered, but in this case that is actually not needed. It could be needed at a later stage if an admin is allowed to update all of the roles of a user at once, but that might be overkill for now.

What do you think?

Reviewer: `change/2` is actually meant for internal use, so there's not really a need to perform validations in theory, since you're fully in control of what you pass along with it.

> Wraps the given data in a changeset or adds changes to a changeset.
>
> changes is a map or keyword where the key is an atom representing a field, association or embed and the value is a term. Note the value is directly stored in the changeset with no validation whatsoever. For this reason, this function is meant for working with data internal to the application.

See https://hexdocs.pm/ecto/Ecto.Changeset.html#change/2.

However you can still validate if you want, for example like this:

```ex
def student_role_changeset(user) do
  user
  |> change(%{roles: ["student"]})
  |> validate_roles
end
```

    |> Changeset.validate_inclusion(:roles, ~w(student teacher))
  end
end
```

### Testing

> In the first blog post I didn;t mention testing at all, but it is a very important part of building a reliable app that is easy to refactor and change with confidence.

Reviewer: ```suggestion
In the first blog post I didn't mention testing at all, but it is a very important part of building a reliable app that is easy to refactor and change with confidence.
```

Without going into too much details about a proper and full-fledged test setup (I'll cover that in a separate post in the future), I would like to test this
user role addition just to check if the above holds up and we are actually able to add these roles to a user. Now, when creating this app and setting this up in the
first post I briefly mentioned that I was going to add the user under the accounts context. Now, we currently don't have any contexts setup, so let's take advantage

> of Phoenix generators to set it up after we've alread created the user. I have confidence that Dan has taken care of the basic testing of the basic user level

Reviewer: Who's `Dan`? This whole section is perhaps a little hard to follow if you haven't (recently) read the previous post. Could you maybe reword it to make it easier to follow for people who haven't recently read the previous post?

Author: good point I have reworded and also linked to my previous post while I was at it :)

Now this gives us a starting point to start testing our `accounts` context. We'll actually remove some of the boilerplate tests as we don't need them right now and
keep a subset we want to test now. This is what is left of the tests:
```elixir
defmodule StudentManager.AccountsTest do
  use StudentManager.DataCase

  alias StudentManager.Accounts

  describe "users" do
    alias StudentManager.Accounts.User

    @valid_attrs %{email: "email@example.com", password: "supersecretpassword", confirm_password: "supersecretpassword"}
    @invalid_attrs %{email: "bademail@bad", password: "short", confirm_password: "shot" }

    def user_fixture(attrs \\ %{}) do
      {:ok, user} =
        attrs
        |> Enum.into(@valid_attrs)
        |> Accounts.create_user()

      user
    end

    test "create_user/1 with valid data creates a user" do
      assert {:ok, %User{} = user} = Accounts.create_user(@valid_attrs)
    end

    test "create_user/1 with invalid data returns error changeset" do
      assert {:error, %Ecto.Changeset{}} = Accounts.create_user(@invalid_attrs)
    end

    test "create_user/1 creates a user with a default role of student" do

>       {:ok, user} = Accounts.create_user(@valid_attrs)

Reviewer: I'm not sure this belongs here? It's not the responsibility of the `Accounts` module, but the `Account` module, right?

Author: Well it is not the `Accounts module`, but rather the `Accounts context` that is the API layer responsible for any interaction related to accounts. Inside the Accounts context we have access to users, parents, invoicing etc. Maybe not precisely these, but that is the idea behind this. The Accounts context is actually what they use in the Phoenix book.

      assert user.roles == ["student"]
    end

    test "create_user/1 validates a correct role" do
      assert {:error, %Ecto.Changeset{}} = Accounts.create_user(Map.put(@valid_attrs, :roles, ["pilot", "teacher"]))
    end
  end
end
```

> Now, we could go ahead and add ExMachina (the elixir equivalent of FactoryBot, if you come from ruby...), but for now, let's not :)

Reviewer: ```suggestion
Now, we could go ahead and add ExMachina (the Elixir equivalent of FactoryBot, if you come from Ruby...), but for now, let's not :)
```

Implementing a Plug is very straightforward and if you want to read up on this, make sure to visit the Plug docs, they're awesome! For now we'll use the suggested implementation
that Dan provides in his guide and leave it at that for now. We'll definitely expand and refine it when we discover places in our app that need some more custom control, but for now I
think I have drained your mental power enough for the day and leave with this Plug:

```elixir
defmodule StudentManagerWeb.Plugs.AuthorizationPlug do
  @moduledoc """
  This plug ensures that a user has a particular role.

  ## Example

      plug StudentManagerWeb.Plugs.AuthorizationPlug, [:student, :teacher]

      plug StudentManagerWeb.Plugs.AuthorizationPlug, :teacher

      plug StudentManagerWeb.Plugs.AuthorizationPlug, ~w(student teacher)a
  """
  alias StudentManagerWeb.Router.Helpers, as: Routes
  alias Phoenix.Controller
  alias Plug.Conn
  alias Pow.Plug

  @doc false
  @spec init(any()) :: any()
  def init(config), do: config

  @doc false
  @spec call(Conn.t(), atom()) :: Conn.t()
  def call(conn, roles) do
    conn
    |> Plug.current_user()
    |> has_role?(roles)
    |> maybe_halt(conn)
  end

>   defp has_role?(nil, _roles), do: false

Reviewer: How about moving this to the `User` module? It's not specific to this plug.

Author: yes that is a possiblity if you are calling it on the user directly. but in this case it is possible to add the Plug to a pipeline and I think then it does make sense here, right? Or would it make more sense to define it in a user context and import that function here?
